src/main/scala/is/hail/methods/Aggregators.scala

```diff
@@ -17,6 +17,75 @@ import scala.reflect.ClassTag
 
 object Aggregators {
 
+  def buildVariantAggregationsByKey(vsm: VariantSampleMatrix, nKeys: Int, keyMap: Array[Int], ec: EvalContext): Option[(RegionValue) => Iterator[() => Unit]] =
+    buildVariantAggregationsByKey(vsm.sparkContext, vsm.matrixType, vsm.value.localValue, nKeys, keyMap, ec)
+
+  // keyMap is just a mapping of sampleIds.map { s => newKey(s) }
+  def buildVariantAggregationsByKey(sc: SparkContext,
+    typ: MatrixType,
+    localValue: VSMLocalValue,
+    nKeys: Int,
+    keyMap: Array[Int],
+    ec: EvalContext): Option[(RegionValue) => Iterator[() => Unit]] = {
+
+    val aggregations = ec.aggregations
+    if (aggregations.isEmpty)
+      return None
+
+    val localA = ec.a
+    val localNSamples = localValue.nSamples
+    val localSamplesBc = sc.broadcast(localValue.sampleIds)
+    val localAnnotationsBc = sc.broadcast(localValue.sampleAnnotations)
+    val localGlobalAnnotations = localValue.globalAnnotation
+    val localRowType = typ.rowType
+
+    Some({ (rv: RegionValue) =>
+      val ur = new UnsafeRow(localRowType, rv.region, rv.offset)
+
+      val v = ur.get(1)
+      val va = ur.get(2)
+      val gs = ur.getAs[IndexedSeq[Annotation]](3)
+
+      val aggs = MultiArray2.fill[Aggregator](nKeys, aggregations.size)(null)
+      var nk = 0
+      while (nk < nKeys) {
+        var nagg = 0
+        while (nagg < aggregations.size) {
+          aggs.update(nk, nagg, aggregations(nagg)._3.copy())
+          nagg += 1
+        }
+        nk += 1
+      }
+      localA(0) = localGlobalAnnotations
+      localA(1) = v
+      localA(2) = va
+
+      var i = 0
+      val git = gs.iterator
```

Comment:
don't use the iterator here -- gs is an UnsafeIndexedSeq, so you can index into it, which will be lower overhead.